ctapipe/image/modifications.py

```diff
@@ -0,0 +1,189 @@
+from numba import njit
+from ..core.component import TelescopeComponent
+from ..core.traits import FloatTelescopeParameter, BoolTelescopeParameter, Int
+from ..instrument import SubarrayDescription
+
+
+__all__ = [
+    "ImageModifier",
+]
+
+
+def _add_noise(image, noise_level, rng=None, correct_bias=True):
+    """
+    Create a new image with added poissonian noise
+    """
+    if not rng:
+        rng = np.random.default_rng()
+
+    noisy_image = image.copy()
+    noise = rng.poisson(noise_level, size=image.shape)
+    noisy_image += noise
+
+    if correct_bias:
+        noisy_image -= noise_level
+
+    return noisy_image
+
+
+@njit(cache=True)
+def _smear_psf_randomly(
+    image, fraction, indices, indptr, smear_probabilities, seed=None
+):
+    """
+    Create a new image with values smeared across the
+    neighbor pixels specified by `indices` and `indptr`.
+    These are what defines the sparse neighbor matrix
+    and are available as attributes from the neighbor matrix.
+    The amount of charge that is distributed away from a given
+    pixel is drawn from a poissonian distribution.
+    The distribution of this charge among the neighboring
+    pixels follows a multinomial.
+    Pixels at the camera edge lose charge this way.
+    No geometry is available in this function due to the numba optimization,
+    so the indices, indptr and smear_probabilities have to match
+    to get sensible results.
+
+    Parameters:
+    -----------
+    image: ndarray
+        1d array of the pixel charge values
+    fraction: float
+        fraction of charge that will be distributed among neighbors (modulo poissonian)
+    indices: ndarray[int]
+        CSR format index array of the neighbor matrix
+    indptr: ndarray[int]
+        CSR format index pointer array of the neighbor matrix
+    smear_probabilities: ndarray[float]
+        shape: (n_neighbors, )
+        A priori distribution of the charge amongst neighbors.
+        In most cases probably of the form np.full(n_neighbors, 1/n_neighbors)
+    seed: int
+        Random seed for the numpy rng.
+        Because this is numba optimized, a rng instance can not be used here
+
+    Returns:
+    --------
+    new_image: ndarray
+        1d array with smeared values
+    """
+    new_image = image.copy()
+    if seed is not None:
+        np.random.seed(seed)
+
+    for pixel in range(len(image)):
+
+        if image[pixel] <= 0:
+            continue
+
+        to_smear = np.random.poisson(image[pixel] * fraction)
+        if to_smear == 0:
+            continue
+
+        # remove light from current pixel
+        new_image[pixel] -= to_smear
+
+        # add light to neighbor pixels
+        neighbors = indices[indptr[pixel] : indptr[pixel + 1]]
+        n_neighbors = len(neighbors)
+
+        # all neighbors are equally likely to receive the charge
+        # we always distribute the charge into 6 neighbors, so that charge
+        # on the edges of the camera is lost
```

Comment:
Is this comment valid? don't we smear into `len(smear_probabilities)` neighbors?

Reply:
They do, thats why you lose something
So you split the charge into 6 (for hexagonal pixels) parts, but only add it to existing neighbors.
Say you draw
```
[1,1,3,1,2,3]
```
and have 4 neighbors (because you are at the edge of the camera). Then this
https://github.com/cta-observatory/ctapipe/blob/63ea86ee52ecfd8969c124697d536854dc02b8b7/ctapipe/image/modifications.py#L96-L98

 stops after 4 iterations and you lose
`2+3 = 5` charge

Reply:
ok, got it